libs/core-ui/src/lib/util/calculateAUCData.ts

```diff
@@ -0,0 +1,160 @@
+    let truePositives = 0;
+    let falsePositives = 0;
+    let trueNegatives = 0;
+    let falseNegatives = 0;
+    for (const [index, yProba] of probabilityY.entries()) {
+      // if the probability of predicting the positive label is greater than the
+      // threshold then it's a true positive.
+      // otherwise, it's a false positive
+      if (yProba < threshold) {
+        if (binY[index]) {
+          falseNegatives++;
+        } else {
+          trueNegatives++;
+        }
+      } else if (binY[index]) {
+        truePositives++;
+      } else {
+        falsePositives++;
+      }
+    }
+    addROCPoint(
+      truePositives,
+      falsePositives,
+      trueNegatives,
+      falseNegatives,
+      points
+    );
+  }
+  points.push({ x: 0, y: 0 });
+
+  // construct the convex hull of the ROC points
+  const sortedPoints = orderBy(points, ["x", "y"], ["asc", "desc"]);
+  const convexHullPoints = [];
+  let highestY = -Infinity;
+  for (let i = 0; i < sortedPoints.length; i++) {
+    // only add points which have a greater y value than the last seen y
+    if (sortedPoints[i].y >= highestY) {
+      convexHullPoints.push(sortedPoints[i]);
+      highestY = sortedPoints[i].y;
+    }
+    let frontIterator = i + 1;
+    while (
+      frontIterator < sortedPoints.length &&
+      sortedPoints[i].x === sortedPoints[frontIterator].x
+    ) {
+      frontIterator++;
+    }
+    i = frontIterator;
+  }
+  const rocData: IROCData = {
+    points: convexHullPoints
+  };
+  return rocData;
+}
+
+function addROCPoint(
+  truePositives: number,
+  falsePositives: number,
+  trueNegatives: number,
+  falseNegatives: number,
+  points: IPoint[]
+): void {
+  // prevent division by 0
+  const totalNegatives = trueNegatives + falsePositives;
+  const totalPositives = truePositives + falseNegatives;
+  const tpr = totalPositives === 0 ? 1 : truePositives / totalPositives;
+  const fpr = totalNegatives === 0 ? 1 : falsePositives / totalNegatives;
+  points.push({ x: fpr, y: tpr });
+}
+
+export function binarizeData(
+  yData: string[] | number[] | number[][],
+  classes: string[] | number[]
+): number[][] {
+  // binarize labels in a one-vs-all fashion
+  const yBinData: number[][] = [];
+  for (const yDatum of yData) {
+    const binaryData = classes.map((c) => {
+      return c === yDatum ? 1 : 0;
+    });
+    yBinData.push(binaryData);
+  }
+  return yBinData;
+}
+
+// based on https://msdata.visualstudio.com/Vienna/_git/AzureMlCli?path=/src/azureml-metrics/azureml/metrics/_classification.py&version=GBmaster
+export function calculateAUCData(dataset: IDataset): SeriesOptionsType[] {
+  if (!dataset.probability_y || !dataset.class_names) {
+    // TODO: show warning message ?
+    return [...getStaticROCData()];
+  }
+
+  // TODO: for some reason in the adult census data, classnames doesn't match the true_y data
+  // need to confirm how this data aligns. temporarily hard code classnames to 0 and 1 for now
+  const cNames = [0, 1];
```

Comment:
This should only run for binary classification, right? So we probably need a check somewhere to disable the component otherwise.

For multiclass one could do one vs all but I don't know if anyone wants that.

Reply:
in the studio, there's an auc chart for multiclass so i assumed we'd want that (binarizeData is supposed to handle this case)

Reply:
That's probably 1 vs all, so that makes sense.